core/templates/domain/topic/topic-rights-backend-api.service.ts

```diff
@@ -16,162 +16,169 @@
-       * specified topic ID with a new topic rights object.
-       */
-      cacheTopicRights: function(topicId, topicRights) {
-        topicRightsCache[topicId] = angular.copy(topicRights);
-      },
-
-      /**
-       * Publishes a topic.
-       */
-      publishTopic: function(topicId) {
-        return $q(function(resolve, reject) {
-          _setTopicStatus(topicId, true, resolve, reject);
-        });
-      },
-
-      sendMail: function(topicId, topicName) {
-        return $q(function(resolve, reject) {
-          _sendMail(topicId, topicName, resolve, reject);
-        });
-      },
-
-      /**
-       * Unpublishes a topic.
-       */
-      unpublishTopic: function(topicId) {
-        return $q(function(resolve, reject) {
-          _setTopicStatus(topicId, false, resolve, reject);
-        });
+    this.httpClient.put(sendMailUrl, putParams).toPromise().then((response) => {
+      successCallback();
+    }, (errorResponse) => {
+      errorCallback(errorResponse.error.error);
+    });
+  }
+
+  private _isCached(topicId: string): boolean {
+    return this.topicRightsCache.hasOwnProperty(topicId);
+  }
+
+  fetchTopicRights(topicId: string): Promise<TopicRightsBackendResponse> {
+    return new Promise((resolve, reject) => {
+      this._fetchTopicRights(topicId, resolve, reject);
+    });
+  }
+
+  /**
+   * Behaves exactly as fetchTopicRights (including callback
+   * behavior and returning a promise object), except this function will
+   * attempt to see whether the given topic rights has been
+   * cached. If it has not yet been cached, it will fetch the topic
+   * rights from the backend. If it successfully retrieves the topic
+   * rights from the backend, it will store it in the cache to avoid
+   * requests from the backend in further function calls.
+   */
+  loadTopicRights(topicId: string): Promise<TopicRightsBackendResponse> {
+    return new Promise((resolve, reject) => {
+      if (this._isCached(topicId)) {
+        resolve(this.topicRightsCache[topicId]);
+      } else {
+        this._fetchTopicRights(topicId, (topicRights) => {
+          // Save the fetched topic rights to avoid future fetches.
+          this.topicRightsCache[topicId] = topicRights;
+          resolve(this.topicRightsCache[topicId]);
+        }, reject);
       }
-    };
+    });
+  }
+
+  /**
+   * Returns whether the given topic rights is stored within the
+   * local data cache or if it needs to be retrieved from the backend
+   * upon a laod.
+   */
+  isCached(topicId: string): boolean {
+    return this._isCached(topicId);
+  }
+
+  /**
+   * Replaces the current topic rights in the cache given by the
+   * specified topic ID with a new topic rights object.
+   */
+  cacheTopicRights(
+      topicId: string,
+      topicRights: TopicRightsBackendResponse): void {
+    this.topicRightsCache[topicId] = topicRights;
+  }
+
+  /**
+   * Publishes a topic.
+   */
+  publishTopic(topicId: string): Promise<TopicRightsBackendResponse> {
```

Comment:
qq: Wouldn't you have to add $rootScope.$applyAsync in places where this function is used?

Reply:
done! there was one file using this function. Added rootScope.applyAsync there.